File: integrations/openbb-agent/optionsahoy_openbb_agent/tools.py
# AlphaLatitude Inc. (c) 2026
"""Tool registry mapping equity-compensation questions to OptionsAhoy calculators.

Each entry pairs an OpenAI function-calling schema (so a language model can extract
the structured inputs from a natural-language question) with the name of the
``OptionsAhoyClient`` method that computes the answer. The agent never reimplements
the math: it forwards the extracted arguments to the keyless OptionsAhoy REST API
through the client and streams the parsed result back.

What each tool computes (not how):

- amt_iso: the multi-year incentive stock option (ISO) exercise schedule that
  minimizes lifetime cost under the alternative minimum tax (AMT).
- nso: the tax and after-tax proceeds of exercising non-qualified stock options
  (NSOs) and holding versus selling.
- rsu_sell_vs_hold: selling vested restricted stock units (RSUs) at vest versus
  holding, compared on an after-tax, risk-adjusted basis.
- concentration: the after-tax cost of diversifying a concentrated single-stock
  position.
- protective_put: the price of three hedge structures (a protective put, a
  zero-cost collar, and a put spread) at a chosen downside protection level and
  tenor.
- qsbs: qualified small business stock (QSBS) eligibility and the resulting
  federal and state capital-gains exclusion.
- equity_funding: which equity lots to sell, and when, to fund a cash goal by a
  target date at the least after-tax cost.
"""

from __future__ import annotations

from typing import Any, Callable, Dict, List

from optionsahoy import OptionsAhoyClient

# Shared enum-ish fields reused across several schemas.
_FILING_STATUS = {
    "type": "string",
    "description": "Tax filing status.",
    "enum": ["single", "married-joint", "married-separate", "head-of-household"],
}
_STATE_CODE = {
    "type": "string",
    "description": "Two-letter state code of tax residence, for example 'CA' or 'NY'.",
}


def _method(client: OptionsAhoyClient, name: str) -> Callable[..., Dict[str, Any]]:
    """Return the bound ``OptionsAhoyClient`` method for a registered tool name."""
    return getattr(client, name)


# Ordered registry. ``parameters`` is a JSON Schema object passed verbatim to the
# language model as an OpenAI function-calling tool definition. ``required`` lists
# only the fields the OptionsAhoy endpoint cannot default; optional modelling
# assumptions (growth, volatility) are left for the API to fill in.
TOOLS: List[Dict[str, Any]] = [
    {
        "name": "amt_iso",
        "description": (
            "Optimize a multi-year incentive stock option (ISO) exercise schedule "
            "under the alternative minimum tax (AMT)."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "shares": {"type": "integer", "description": "Number of ISO shares."},
                "strike": {"type": "number", "description": "Per-share strike price."},
                "fmv": {"type": "number", "description": "Current fair market value per share."},
                "filingStatus": _FILING_STATUS,
                "ordinaryIncome": {"type": "number", "description": "Annual ordinary income."},
                "stateCode": _STATE_CODE,
                "carryforwardCredit": {
                    "type": "number",
                    "description": "Existing AMT credit carryforward, or 0.",
                },
                "horizon": {"type": "integer", "description": "Planning horizon in years."},
                "cashReturnRate": {
                    "type": "number",
                    "description": "Assumed annual return on cash, as a decimal such as 0.04.",
                },
                "grantDate": {"type": "string", "description": "ISO grant date, YYYY-MM-DD."},
                "hasLeftCompany": {
                    "type": "boolean",
                    "description": "Whether the holder has left the company.",
                },
                "terminationDate": {
                    "type": ["string", "null"],
                    "description": "Termination date YYYY-MM-DD, or null if still employed.",
                },
            },
            "required": [
                "shares",
                "strike",
                "fmv",
                "filingStatus",
                "ordinaryIncome",
                "stateCode",
                "carryforwardCredit",
                "horizon",
                "cashReturnRate",
                "grantDate",
                "hasLeftCompany",
                "terminationDate",
            ],
        },
    },
    {
        "name": "nso",
        "description": (
            "Compute the tax and after-tax proceeds of exercising non-qualified stock "
            "options (NSOs) and holding versus selling."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "shares": {"type": "integer", "description": "Number of NSO shares."},
                "strike": {"type": "number", "description": "Per-share strike price."},
                "currentPrice": {"type": "number", "description": "Current share price."},
                "ordinaryIncome": {"type": "number", "description": "Annual ordinary income."},
                "filingStatus": _FILING_STATUS,
                "stateCode": _STATE_CODE,
                "stillEmployed": {
                    "type": "boolean",
                    "description": "Whether the holder is still employed.",
                },
                "holdYears": {
                    "type": "number",
                    "description": "Years to hold after exercise before selling.",
                },
                "holdFunding": {
                    "type": "string",
                    "description": "How the exercise cost is funded.",
                    "enum": ["cash", "cashless", "sell-to-cover"],
                },
            },
            "required": [
                "shares",
                "strike",
                "currentPrice",
                "ordinaryIncome",
                "filingStatus",
                "stateCode",
                "stillEmployed",
                "holdYears",
                "holdFunding",
            ],
        },
    },
    {
        "name": "rsu_sell_vs_hold",
        "description": (
            "Compare selling vested restricted stock units (RSUs) at vest against "
            "holding them, on an after-tax, risk-adjusted basis."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "shares": {"type": "integer", "description": "Number of vested RSU shares."},
                "currentPrice": {"type": "number", "description": "Current share price."},
                "ordinaryIncome": {"type": "number", "description": "Annual ordinary income."},
                "filingStatus": _FILING_STATUS,
                "stateCode": _STATE_CODE,
                "stillEmployed": {
                    "type": "boolean",
                    "description": "Whether the holder is still employed.",
                },
                "holdYears": {
                    "type": "number",
                    "description": "Years to hold the shares before selling.",
                },
            },
            "required": [
                "shares",
                "currentPrice",
                "ordinaryIncome",
                "filingStatus",
                "stateCode",
                "stillEmployed",
                "holdYears",
            ],
        },
    },
    {
        "name": "concentration",
        "description": (
            "Analyze a concentrated single-stock position and the after-tax cost of "
            "diversifying it."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "positionValue": {"type": "number", "description": "Current market value of the position."},
                "costBasis": {"type": "number", "description": "Total cost basis of the position."},
                "acquisitionDate": {"type": "string", "description": "Acquisition date, YYYY-MM-DD."},
                "sector": {"type": "string", "description": "Sector of the holding, for example 'tech'."},
                "stateCode": _STATE_CODE,
                "filingStatus": _FILING_STATUS,
                "ordinaryIncome": {"type": "number", "description": "Annual ordinary income."},
                "totalAssets": {
                    "type": "number",
                    "description": "Total investable assets, to size the concentration.",
                },
            },
            "required": [
                "positionValue",
                "costBasis",
                "acquisitionDate",
                "sector",
                "stateCode",
                "filingStatus",
                "ordinaryIncome",
                "totalAssets",
            ],
        },
    },
    {
        "name": "protective_put",
        "description": (
            "Price a protective put, a zero-cost collar, and a put spread for a stock "
            "position at a given downside protection level and tenor."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "positionValue": {"type": "number", "description": "Market value of the position to hedge."},
                "sector": {"type": "string", "description": "Sector of the holding, for example 'tech'."},
                "protectionLevel": {
                    "type": "number",
                    "description": "Floor as a fraction of current value, for example 0.9 for a 10 percent floor.",
                },
                "tenorYears": {"type": "number", "description": "Hedge tenor in years."},
                "spreadRiskLevel": {
                    "type": "number",
                    "minimum": 0.01,
                    "maximum": 0.2,
                    "description": (
                        "Put spread floor breach risk: probability the stock ends "
                        "below the spread's short strike (presets 0.20 / 0.10 / 0.05 "
                        "/ 0.01, i.e. 1 in 5 / 10 / 20 / 100; off-preset snaps to "
                        "nearest). Affects only the putSpread block. Default 0.10."
                    ),
                },
            },
            "required": ["positionValue", "sector", "protectionLevel", "tenorYears"],
        },
    },
    {
        "name": "qsbs",
        "description": (
            "Check qualified small business stock (QSBS) eligibility and the resulting "
            "federal and state capital-gains exclusion."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "acquisitionDate": {"type": "string", "description": "Share acquisition date, YYYY-MM-DD."},
                "saleDate": {"type": "string", "description": "Planned or actual sale date, YYYY-MM-DD."},
                "entityType": {"type": "string", "description": "Issuer entity type, for example 'us-c-corp'."},
                "acquisitionMethod": {
                    "type": "string",
                    "description": "How the stock was acquired, for example 'original-issuance'.",
                },
                "assetCategory": {
                    "type": "string",
                    "description": "Gross-asset category at issuance, for example 'under-50m'.",
                },
                "industry": {"type": "string", "description": "Issuer industry, for example 'tech-software'."},
                "activeBusiness": {
                    "type": "string",
                    "description": "Whether the issuer met the active-business test, 'yes' or 'no'.",
                },
                "adjustedBasis": {"type": "number", "description": "Adjusted cost basis of the stock."},
                "expectedGain": {"type": "number", "description": "Expected gain on sale."},
                "stateCode": _STATE_CODE,
                "ordinaryIncome": {"type": "number", "description": "Annual ordinary income."},
                "filingStatus": _FILING_STATUS,
            },
            "required": [
                "acquisitionDate",
                "saleDate",
                "entityType",
                "acquisitionMethod",
                "assetCategory",
                "industry",
                "activeBusiness",
                "adjustedBasis",
                "expectedGain",
                "stateCode",
                "ordinaryIncome",
                "filingStatus",
            ],
        },
    },
    {
        "name": "equity_funding",
        "description": (
            "Plan which equity lots to sell, and when, to fund a cash goal by a target "
            "date with the least after-tax cost."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "targetAfterTax": {
                    "type": "number",
                    "description": "After-tax cash amount needed.",
                },
                "targetDate": {"type": "string", "description": "Date the cash is needed by, YYYY-MM-DD."},
                "ordinaryIncome": {"type": "number", "description": "Annual ordinary income."},
                "filingStatus": _FILING_STATUS,
                "stateCode": _STATE_CODE,
                "currentPrice": {
                    "type": "number",
                    "description": "Current share price, when using the simple single-position form.",
                },
            },
            "required": [
                "targetAfterTax",
                "targetDate",
                "ordinaryIncome",
                "filingStatus",
                "stateCode",
            ],
        },
    },
]

# Fast lookup by tool name.
TOOLS_BY_NAME: Dict[str, Dict[str, Any]] = {tool["name"]: tool for tool in TOOLS}


def openai_tool_specs() -> List[Dict[str, Any]]:
    """Return the registry as OpenAI ``tools`` (function-calling) specifications."""
    return [
        {
            "type": "function",
            "function": {
                "name": tool["name"],
                "description": tool["description"],
                "parameters": tool["parameters"],
            },
        }
        for tool in TOOLS
    ]


def call_tool(
    client: OptionsAhoyClient, name: str, arguments: Dict[str, Any]
) -> Dict[str, Any]:
    """Dispatch ``name`` to the matching ``OptionsAhoyClient`` method.

    Raises ``KeyError`` for an unknown tool name. Arguments are forwarded as
    keyword arguments; the client validates and posts them to the keyless API.
    """
    if name not in TOOLS_BY_NAME:
        raise KeyError(f"Unknown OptionsAhoy tool: {name}")
    return _method(client, name)(**arguments)

File: integrations/js/optionsahoy-ai-sdk/src/tools.ts
// AlphaLatitude Inc. © 2026
//
// Vercel AI SDK tool definitions, one per OptionsAhoy calculator endpoint.
// Each tool carries a description, a zod `parameters` schema, and an `execute`
// that POSTs to the keyless REST endpoint and returns the parsed `result`.
//
// Descriptions say what each tool computes, not how, and deabbreviate acronyms
// on first use. Results are independent calculations; integrated multi-year,
// multi-position optimization is available in the OptionsAhoy beta.

import { tool } from 'ai';
import type { z } from 'zod';

import { callEndpoint, type EndpointSlug, type OptionsAhoyClientOptions } from './client.js';
import {
  amtIsoParameters,
  concentrationParameters,
  equityFundingParameters,
  nsoParameters,
  protectivePutParameters,
  qsbsParameters,
  rsuParameters,
} from './schemas.js';

/** Options accepted by the tool factory (base URL and/or a custom fetch). */
export type { OptionsAhoyClientOptions } from './client.js';

/** Build an `execute` that POSTs the validated args to `slug` and returns `result`. */
function makeExecute<S extends z.ZodTypeAny>(slug: EndpointSlug, options: OptionsAhoyClientOptions) {
  return async (args: z.infer<S>): Promise<unknown> =>
    callEndpoint(slug, args as Record<string, unknown>, options);
}

/**
 * Return the seven OptionsAhoy calculators as Vercel AI SDK tools, keyed by
 * their canonical tool name. Spread the result into `generateText`/`streamText`
 * `tools`, or pick individual tools.
 *
 * @param options Optional `baseURL` and `fetch` overrides. Defaults are keyless
 *   and point at https://optionsahoy.com.
 */
export function createOptionsAhoyTools(options: OptionsAhoyClientOptions = {}) {
  return {
    amt_iso_optimize: tool({
      description:
        'Optimize a multi-year incentive stock option (ISO) exercise schedule under the alternative minimum tax (AMT). Returns how many shares to exercise each year to maximize after-tax net final value across the planning horizon, with federal and state tax for the lump-sum, even-split, and optimized paths. Use for ISO planning; for NSOs use nso_calculate, for RSUs use rsu_sell_vs_hold.',
      parameters: amtIsoParameters,
      execute: makeExecute<typeof amtIsoParameters>('amt-iso', options),
    }),

    nso_calculate: tool({
      description:
        'Calculate the tax and after-tax proceeds of exercising non-qualified stock options (NSOs), comparing selling at exercise against holding for later long-term capital gains. Covers federal, state, and FICA (Social Security and Medicare) tax at exercise and capital-gains treatment at sale.',
      parameters: nsoParameters,
      execute: makeExecute<typeof nsoParameters>('nso', options),
    }),

    rsu_sell_vs_hold: tool({
      description:
        'Compare selling vested restricted stock units (RSUs) at vest against holding them, on an after-tax, risk-adjusted basis, and return which choice is expected to leave more wealth. Covers ordinary tax and withholding at vest and capital-gains treatment on later sale.',
      parameters: rsuParameters,
      execute: makeExecute<typeof rsuParameters>('rsu-sell-vs-hold', options),
    }),

    concentration_analyze: tool({
      description:
        'Analyze a concentrated single-stock position: its share of total assets, drawdown exposure at 30/50/70 percent drops, and the after-tax cost of three responses (sell down to target weight, hold, or hedge with a put or zero-cost collar) over a three-year horizon.',
      parameters: concentrationParameters,
      execute: makeExecute<typeof concentrationParameters>('concentration', options),
    }),

    protective_put_price: tool({
      description:
        'Price three hedge structures for a single-stock position at a chosen downside-protection level and tenor: a protective put, a zero-cost collar, and a put spread. Returns the premium, annualized cost, maximum loss, and a recommendation for each.',
      parameters: protectivePutParameters,
      execute: makeExecute<typeof protectivePutParameters>('protective-put', options),
    }),

    qsbs_check: tool({
      description:
        'Check qualified small business stock (QSBS) Section 1202 eligibility across the six statutory tests and compute the resulting federal and state capital-gains exclusion on a planned sale, including the per-issuer and ten-times-basis caps.',
      parameters: qsbsParameters,
      execute: makeExecute<typeof qsbsParameters>('qsbs', options),
    }),

    equity_funding_plan: tool({
      description:
        'Plan which equity lots to sell, and when, to fund a cash goal by a target date at the least after-tax cost, accounting for holding-period thresholds and shortfall risk. Returns four named plans on the risk/wealth frontier (lock in now, balanced, hold for growth, recommended).',
      parameters: equityFundingParameters,
      execute: makeExecute<typeof equityFundingParameters>('equity-funding', options),
    }),
  } as const;
}

/** The canonical tool names exported by {@link createOptionsAhoyTools}. */
export type OptionsAhoyToolName = keyof ReturnType<typeof createOptionsAhoyTools>;

/** Keyless, default-configured tool set pointing at https://optionsahoy.com. */
export const optionsAhoyTools = createOptionsAhoyTools();
